src/aof.c

```diff
@@ -1545,19 +1504,23 @@ int loadAppendOnlyFiles(aofManifest *am) {
     off_t total_size = 0;
     sds aof_name;
 
-    /* If the 'server.aof_filename' directory does not exist, we may be starting 
-     * from an old redis version. so we must fall back to the previous loading mode. */
-    if (!dirExists(server.aof_filename)) {
-        serverAssert(am->base_aof_info == NULL && listLength(am->incr_aof_list) == 0);
-        sds tmp_dirname = getTempAofDirName();
-        if (fileExist(server.aof_filename) || dirExists(tmp_dirname)) {
-            if (aofUpgradePrepare(am) == C_ERR) {
-                sdsfree(tmp_dirname);
-                return AOF_FAILED;
-            }
-        }
-        sdsfree(tmp_dirname);
-    }
+    /* If the 'server.aof_filename' file exists in dir, we may be starting 
+     * from an old redis version. We will use enter upgrade mode in three situations.
+     *  
+     * 1. If the 'server.aof_dirname' directory not exist
+     * 2. If the 'server.aof_dirname' directory exists but is empty
+     * 3. If the 'server.aof_dirname' directory exists and the manifest file it contains 
+     *    has only one base AOF record, and the file name of this base AOF is 'server.aof_filename',
+     *    and the 'server.aof_filename' file not exist in 'server.aof_dirname' directory
+     * */
+    if (fileExist(server.aof_filename)) {
+        if (!dirExists(server.aof_dirname) || 
+            (am->base_aof_info == NULL && listLength(am->incr_aof_list) == 0) ||
+            (am->base_aof_info != NULL && listLength(am->incr_aof_list) == 0 && 
+            !strcmp(am->base_aof_info->file_name, server.aof_filename) && !aofFileExist(server.aof_filename))) {
```

Comment:
indentation and line break
```suggestion
             !strcmp(am->base_aof_info->file_name, server.aof_filename) && !aofFileExist(server.aof_filename)))
        {
```

Reply:
Fixed.

Reply:
i think the indentation change is in some way missing (because of the parenthesis)